### PR TITLE
Further additions, improvements and fixes to the alpha Java API 

### PR DESCRIPTION
**Pull request by danmason**

### CODE REVIEW

File: crux-core/src/crux/api/alpha/CruxNode.java [hunk 1]
```diff
@@ -93,17 +61,42 @@ private Document document(Object contentHash) {
...
-    public Iterable<Document> history(CruxId id) {
+    public Iterable<EntityTx> history(CruxId id) {
         List<Map<Keyword,Object>> history = node.history(id.toEdn());
         return history.stream()
-            .map(this::document)
+            .map(entity -> EntityTx.entityTx(entity))
             .collect(Collectors.toList());
     }
 
+    /**
```
Comment (jarohen) [inline crux-core/src/crux/api/alpha/CruxNode.java]: @danmason - can you omit these two (`hasSubmittedTx*`) for the time being? not convinced they're going to remain supported

File: crux-core/src/crux/api/alpha/NodeStatus.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    public final String version;
+    public final String revision;
+    public final String kvStore;
+    public final int estimatedNumberOfKeys;
+    public final long size;
+    public final long indexVersion;
+
+    @Override
+    public String toString() {
+        return "NodeStatus{" +
```
Comment (jarohen) [inline crux-core/src/crux/api/alpha/NodeStatus.java:17]: IntelliJ'll generate you a String.format based toString if you ask it to, I think? preferable to string concatenation for both performance and (more subjective) readability

File: crux-core/src/crux/api/alpha/NodeStatus.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package crux.api.alpha;
+
+import clojure.lang.Keyword;
+
+import java.util.Map;
+
+public class NodeStatus {
+    public final String version;
```
Comment (jarohen) [inline crux-core/src/crux/api/alpha/NodeStatus.java:8]: we should consider which of these we want to support, and whether we want to support them as a separate object or 'just' values on the node itself, in Java.

maybe not 100% either way, too - version, revision, indexversion etc are all static when the node's created; the others are dynamic

File: crux-core/src/crux/api/alpha/EntityTxWithDocument.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package crux.api.alpha;
+
+import clojure.lang.Keyword;
+
+import java.util.Map;
+
+public class EntityTxWithDocument {
```
Comment (jarohen) [inline crux-core/src/crux/api/alpha/EntityTxWithDocument.java:7]: hmm :thinking:

Comment (danmason) [reply crux-core/src/crux/api/alpha/EntityTxWithDocument.java:7]: Agreed on the hmm - was unsure if I should just edit what I had with EntityTx, or return some sort of tuple (as I did here, sort of)

File: crux-core/src/crux/api/alpha/TxResult.java [hunk 5]
```diff
@@ -1,18 +1,33 @@
...
+
+import static crux.api.alpha.Util.keyword;
 
 public class TxResult {
+    private static final Keyword TX_TIME = keyword("crux.tx/tx-time");
+    private static final Keyword TX_ID = keyword("crux.tx/tx-id");
+
     public final Date txTime;
     public final long txId;
+    private Map<Keyword, ?> txResultMap;
```
Comment (jarohen) [inline crux-core/src/crux/api/alpha/TxResult.java]: nit: `final`

File: crux-core/src/crux/api/alpha/CruxNode.java (hunk 1 above)
Comment (danmason) [reply crux-core/src/crux/api/alpha/CruxNode.java]: Yeah, can do - remove, or make private?

File: crux-core/src/crux/api/alpha/NodeStatus.java (hunk 2 above)
Comment (danmason) [reply crux-core/src/crux/api/alpha/NodeStatus.java:17]: Might need to check that? I think this is what it does by default, so will need to have a look at my options

File: crux-core/src/crux/api/alpha/TransactionOperation.java [hunk 6]
```diff
@@ -1,11 +1,26 @@
...
 public abstract class TransactionOperation {
+    static final Keyword TX_PUT = keyword("crux.tx/put");
+    static final Keyword TX_EVICT = keyword("crux.tx/evict");
+    static final Keyword TX_CAS = keyword("crux.tx/cas");
+    static final Keyword TX_DELETE = keyword("crux.tx/delete");
+
     TransactionOperation() {
     }
 
+    // May want to move these - are inherited by the classes
```
Comment (jarohen) [inline crux-core/src/crux/api/alpha/TransactionOperation.java:23]: the static constructors?

File: crux-core/src/crux/api/alpha/NodeStatus.java (hunk 3 above)
Comment (danmason) [reply crux-core/src/crux/api/alpha/NodeStatus.java:8]: Yeah - this was a sort of placeholder, one of the classes I wanted a bit more thinking on (same as EntityTx)

File: crux-core/src/crux/api/alpha/TxResult.java (hunk 5 above)
Comment (danmason) [reply crux-core/src/crux/api/alpha/TxResult.java]: Yeah - need to check a few places that probably need 'final'

File: crux-core/src/crux/api/alpha/TransactionOperation.java (hunk 6 above)
Comment (danmason) [reply crux-core/src/crux/api/alpha/TransactionOperation.java:23]: Yep, in the generated Javadocs they're listed as inherited functions

**Comment by jarohen** [comment]
@danmason - this is a good PR but I'm going to close it for now because I'd like to sort out the underlying Clojure API before checking the alpha Java API for completeness. when that's done, let's update and re-open this PR

**Comment by danmason** [comment]
:+1: Good by me, better than it sitting open with no movement in our PR list :)

**Pull request closed without merging**